File: packages/node/src/payment.ts
export type XPaymentAuthorization = {
  from: string; // hex address
  to: string;   // hex address (resource server)
  value: string; // decimal string amount
  validAfter: string;  // unix seconds (string)
  validBefore: string; // unix seconds (string)
  nonce: string;       // hex
};

export type XPaymentPayload = {
  signature: string; // hex 0x...
  authorization: XPaymentAuthorization;
};

export type XPaymentHeader = {
  x402Version: 1;
  scheme: string; // e.g., "exact"
  network: string; // e.g., base, base-sepolia
  payload: XPaymentPayload;
};

export type XPaymentResponseHeader = {
  x402Version: 1;
  scheme: string; // e.g., "exact"
  network: string;
  payload: Record<string, any>; // opaque for now; integrity-protected in AAD
};

export type PaymentLike = XPaymentHeader | XPaymentResponseHeader;

export function isXPayment(obj: any): obj is XPaymentHeader {
  return obj && typeof obj === "object" && obj.payload && typeof obj.payload.signature === "string" && obj.authorization === undefined;
}

export function isXPaymentResponse(obj: any): obj is XPaymentResponseHeader {
  return obj && typeof obj === "object" && obj.payload && typeof obj.payload === "object" && obj.payload.signature === undefined;
}

export function normalizePaymentLike<T extends PaymentLike>(input: T): T {
  // Minimal structural checks; leave deep validation to integrators for now
  if (!input || typeof input !== "object") throw new Error("X_PAYMENT_SCHEMA");
  if (input.x402Version !== 1) throw new Error("X_PAYMENT_VERSION");
  if (!input.scheme || typeof input.scheme !== "string") throw new Error("X_PAYMENT_SCHEMA");
  if (!input.network || typeof input.network !== "string") throw new Error("X_PAYMENT_SCHEMA");
  if (!input.payload || typeof input.payload !== "object") throw new Error("X_PAYMENT_SCHEMA");
  return input;
}

function deepCanonicalize(obj: any): any {
  if (obj === null || typeof obj !== "object") return obj;
  if (Array.isArray(obj)) return obj.map(deepCanonicalize);
  const keys = Object.keys(obj).sort();
  const out: any = {};
  for (const k of keys) out[k] = deepCanonicalize(obj[k]);
  return out;
}

export function synthesizePaymentHeaderValue(p: PaymentLike): string {
  // Emit compact, stable JSON string (header value is JSON per x402 examples)
  return JSON.stringify(deepCanonicalize(p));
}

export function parsePaymentHeaderValue(s: string): PaymentLike | null {
  // Accept either raw JSON or base64-encoded JSON (be liberal in what we accept)
  try {
    const j = JSON.parse(s);
    return normalizePaymentLike(j);
  } catch (_) {
    try {
      const dec = Buffer.from(s, "base64").toString("utf8");
      const j2 = JSON.parse(dec);
      return normalizePaymentLike(j2);
    } catch {
      return null;
    }
  }
}

export function deriveExtendedAppFromPayment(p: XPaymentHeader): Record<string, string> {
  const a = p.payload.authorization;
  return {
    paymentScheme: String(p.scheme),
    paymentNetwork: String(p.network),
    paymentFrom: String(a.from),
    paymentTo: String(a.to),
    paymentValue: String(a.value),
    paymentValidAfter: String(a.validAfter),
    paymentValidBefore: String(a.validBefore),
    paymentNonce: String(a.nonce),
    paymentSignature: String(p.payload.signature),
  };
}

File: packages/python/x402_hpke/payment.py
from __future__ import annotations
from typing import Any, Dict, Optional, Tuple, TypedDict
import base64
import json


class XPaymentAuthorization(TypedDict):
    from_: str  # hex address; underscore to avoid keyword
    to: str
    value: str
    validAfter: str
    validBefore: str
    nonce: str


class XPaymentPayload(TypedDict):
    signature: str
    authorization: Dict[str, Any]


class XPaymentHeader(TypedDict):
    x402Version: int
    scheme: str
    network: str
    payload: XPaymentPayload


class XPaymentResponseHeader(TypedDict, total=False):
    x402Version: int
    scheme: str
    network: str
    payload: Dict[str, Any]


PaymentLike = Dict[str, Any]


def _deep_canonicalize(obj: Any) -> Any:
    if obj is None or isinstance(obj, (str, int, float, bool)):
        return obj
    if isinstance(obj, list):
        return [_deep_canonicalize(x) for x in obj]
    if isinstance(obj, dict):
        return {k: _deep_canonicalize(obj[k]) for k in sorted(obj.keys())}
    return obj


def synthesize_payment_header_value(p: PaymentLike) -> str:
    # Emit compact, stable JSON string
    return json.dumps(_deep_canonicalize(p), separators=(",", ":"))


def parse_payment_header_value(s: str) -> Optional[PaymentLike]:
    # Accept raw JSON or base64-encoded JSON
    try:
        j = json.loads(s)
        return normalize_payment_like(j)
    except Exception:
        try:
            dec = base64.b64decode(s + "==").decode("utf-8")
            j2 = json.loads(dec)
            return normalize_payment_like(j2)
        except Exception:
            return None


def normalize_payment_like(p: PaymentLike) -> PaymentLike:
    if not isinstance(p, dict):
        raise ValueError("X_PAYMENT_SCHEMA")
    if p.get("x402Version") != 1:
        raise ValueError("X_PAYMENT_VERSION")
    if not isinstance(p.get("scheme"), str) or not isinstance(p.get("network"), str):
        raise ValueError("X_PAYMENT_SCHEMA")
    payload = p.get("payload")
    if not isinstance(payload, dict):
        raise ValueError("X_PAYMENT_SCHEMA")
    # minimal checks only; leave deep chain-specific validation to integrators
    return p


def derive_extended_app_from_payment(p: PaymentLike) -> Dict[str, str]:
    try:
        auth = p.get("payload", {}).get("authorization", {})
        return {
            "paymentScheme": str(p.get("scheme")),
            "paymentNetwork": str(p.get("network")),
            "paymentFrom": str(auth.get("from")),
            "paymentTo": str(auth.get("to")),
            "paymentValue": str(auth.get("value")),
            "paymentValidAfter": str(auth.get("validAfter")),
            "paymentValidBefore": str(auth.get("validBefore")),
            "paymentNonce": str(auth.get("nonce")),
            "paymentSignature": str(p.get("payload", {}).get("signature")),
        }
    except Exception:
        return {}

